pkg/sessions/dynamodb/client.go

```diff
@@ -0,0 +1,74 @@
+package dynamodb
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	dynamo "github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"time"
+)
+
+type Client struct {
+	*dynamo.DynamoDB
+}
+
+var TableName = "Session"
+
+type dynamoItem struct {
+	SessionKey string
+	Value      string
+}
+
+func newClient(c *dynamo.DynamoDB) *Client {
+	return &Client{DynamoDB: c}
+}
+
+func (c *Client) Get(_ context.Context, key string) ([]byte, error) {
+	result, err := c.GetItem(&dynamo.GetItemInput{
+		TableName: aws.String(TableName),
+		Key: map[string]*dynamo.AttributeValue{
+			"SessionKey": {
+				S: aws.String(key),
+			},
+		},
+	})
+	item := dynamoItem{}
+
+	err = dynamodbattribute.UnmarshalMap(result.Item, &item)
+	if err != nil {
+		// TODO: Return default val?
+		return nil, fmt.Errorf("failed unmarshalling record: %v", err)
+	}
+
+	val, err := b64.StdEncoding.DecodeString(item.Value)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode dynamodb session: %v", err)
+	}
+
+	return val, nil
+}
+
+func (c *Client) Set(ctx context.Context, key string, value []byte, expiration time.Duration) error {
```

Comment:
How will we handle the `expiration` in DynamoDB? Otherwise we run the risk of this session database growing uncontained (only users that explicitly sign out or return after a session expiration will hit the `Del` method).

Reply:
My initial thoughts were to handle it via an explicit check against an property on the object stored in Dynamo. DynamoDB does offer TTL support, which whilst it isn't (documentation says it can take up to 48 hours) we can prevent the data set growing uncontrollably.

Reply:
While the session we'll be unmarshalling has an expiration field inside of it, it looks like any checks on it might be very delayed.

I'll need to dig more, but on first glance: It seems like we rely on `Get`/`Load` in these session stores not returning a session if it is expired? (we likely also have appropriate expirations set on the cookie that has the ticket/key you will be using, so we are fine there as a backup).

So even if it takes a while for the data to actually be deleted (which is fine), we'll need to make sure we don't return data past expiration at the code level.

Reply:
>So even if it takes a while for the data to actually be deleted (which is fine), we'll need to make sure we don't return data past expiration at the code level.

Agreed. So store the expiry as a field and explicitly check it on get, but have a TTL for general clean up to stop the table getting too large. Using the TTL for clean up doesn't count against your RCU (read capacity units) so works out cheaper than deleting the records in the long run)